lib/ansible/modules/cloud/amazon/storage_gateway.py

```diff
@@ -0,0 +1,241 @@
+                GatewayTimezone=params['GatewayTimezone']
+            )
+            disks_response = client.list_local_disks(
+                GatewayARN=result['GatewayARN']
+            )
+            result['gateway_arn'] = response['GatewayARN']
+            result['Disks'] = disks_response['Disks']
+            result['changed'] = True
+            return result
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't update storage gateway")
+
+    return result
+
+
+def delete_gateway(client, module, result):
+    try:
+        response = client.delete_gateway(
+            GatewayARN=result['GatewayARN']
+        )
+        result['changed'] = True
+        return result
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't delete storage gateway")
+
+    return result
+
+
+def main():
+    module = AnsibleAWSModule(
+        argument_spec={
+            'name': dict(type='str', required=True),
+            'state': dict(type='str', choices=['present', 'absent'], default='present'),
+            'activation_key': dict(type='str', required=True),
+            'timezone': dict(type='str', required=True),
+            'gateway_region': dict(type='str', required=True),
+            'gateway_type': dict(type='str', default='STORED', choices=['STORED', 'CACHED', 'VTL', 'FILE_S3']),
+            'tape_drive_type': dict(type='str', choices=['IBM-ULT3580-TD5']),
+            'medium_changer_type': dict(type='str', choices=['STK-L700', 'AWS-Gateway-VTL']),
+        },
+        supports_check_mode=False,
+    )
+
+    result = {
+        'changed': False,
+        'gateway_arn': ''
+    }
+
+    desired_state = module.params.get('state')
+
+    params = {}
+    params['GatewayName'] = module.params.get('name')
+    params['ActivationKey'] = module.params.get('activation_key')
+    params['GatewayTimezone'] = module.params.get('timezone')
+    params['GatewayRegion'] = module.params.get('gateway_region')
+    params['GatewayType'] = module.params.get('gateway_type')
+    if module.params.get('tape_drive_type'):
+        params['TapeDriveType'] = module.params.get('tape_drive_type')
+    if module.params.get('medium_changer_type'):
+        params['MediumChangerType'] = module.params.get('medium_changer_type')
+
+    client = module.client('storagegateway')
+
+    gateway_status = gateway_exists(client, module, params, result)
+
+    if desired_state == 'present':
+        if not gateway_status:
+            create_gateway(client, module, params, result)
+        if gateway_status:
+            update_gateway(client, module, params, result)
+
+    if desired_state == 'absent':
+        if gateway_status:
+            delete_gateway(client, module, result)
+
+    module.exit_json(changed=result['changed'], gateway_arn=result['gateway_arn'])
```

Comment:
Not a blocker, but it seems like returning tuples could be an alternative to passing around a reference to a result dictionary that gets modified and not returned.